File: docs/source/arrays-and-buffers/flatten.c
#include <stddef.h>

void flatten_3D(double * in, double * out, size_t * shape) {
  /* Copy the contents of a contiguous 3D array into a flat array. */
  size_t n = 0;
  for (size_t i = 0; i < shape[0]; i++) {
    for (size_t j = 0; j < shape[1]; j++) {
      for (size_t k = 0; k < shape[2]; k++) {
        // Convert from 3D indices to flat index.
        size_t flat_index = (i * shape[1] + j) * shape[2] + k;
        // Copy the value.
        out[n++] = in[flat_index];
      }
    }
  }
}

// ---

void flatten_strided_3D(double * in, double * out, size_t * shape, size_t * strides) {
  /* Copy the contents of a non-contiguous 3D array into a flat array. */
  size_t n = 0;
  for (size_t i = 0; i < shape[0]; i++) {
    for (size_t j = 0; j < shape[1]; j++) {
      for (size_t k = 0; k < shape[2]; k++) {
        // Convert from 3D strided indices to flat.
        size_t flat_index = (i * strides[0] + j * strides[1] + k * strides[2]);
        // NumPy's stride offsets are in bytes rather than items.
        flat_index /= sizeof(double);
        // Copy the value.
        out[n++] = in[flat_index];
      }
    }
  }
}

// ---

File: docs/source/arrays-and-buffers/flatten.py
# -*- coding: utf-8 -*-
"""
"""

from cslug import CSlug, anchor, ptr, nc_ptr

slug = CSlug(anchor("flatten.c"))
slug.make()

import pytest

pytest.importorskip("numpy")
import numpy as np


def flatten_3D(arr):
    """Wrapper for ``flatten_3D()``."""

    # Normalise `arr`.
    arr = np.asarray(arr, dtype=np.double, order="C")
    # Sanity check that `arr` is 3D.
    assert arr.ndim == 3

    # Create a flat empty output array to populate.
    out = np.empty(arr.size, arr.dtype)

    # Pass `arr`, `out` and the shape of `arr` to the C function.
    # Note the use of `arr.ctypes.shape` which is a ctypes size_t array,
    # instead of `arr.shape` which is a tuple.
    slug.dll.flatten_3D(ptr(arr), ptr(out), ptr(arr.ctypes.shape))

    return out


def flatten_strided_3D(arr):
    """Wrapper for ``flatten_strided_3D()``."""

    # Normalise array dtype but no need to enforce contiguity.
    arr = np.asarray(arr, dtype=np.double)
    assert arr.ndim == 3

    # As before, create a flat empty output array to populate.
    out = np.empty(arr.size, arr.dtype)

    # Pass the arrays, the shape and strides to C.
    # Note, you must use `nc_ptr(arr)` instead of `ptr()` because `arr` is
    # not necessarily contiguous.
    slug.dll.flatten_strided_3D(nc_ptr(arr), ptr(out), ptr(arr.ctypes.shape),
                                ptr(arr.ctypes.strides))
    return out


arr: np.ndarray = np.arange(24, dtype=np.double).reshape((2, 3, 4))
assert np.array_equal(arr.ravel(), flatten_3D(arr))
assert np.array_equal(arr.T.ravel(), flatten_3D(arr.T))
assert np.array_equal(arr.ravel(), flatten_strided_3D(arr))
assert np.array_equal(arr.T.ravel(), flatten_strided_3D(arr.T))
